### PR TITLE
add options to retutn log normalizer & total attention

### PR DESCRIPTION
**Pull request by iclementine**
1. Add a new kernel to compute the total amount of attention that each `k` receives from all `q`s. It returns a `(batch_size, num_heads, seqlen_k)`;
2. Add an option `return_log_normalizer` for `flash_attention`. If `return_log_normalizer` is True, it returns the log normalizer of softmax, a tensor of shape `(batch_size, num_heads, seqlen_q)`;
3. Add an option `return_total_attention` for `flash_attention`. If `return_total_attention` is True, it returns the total attention(the sum of attention weight each `k` receives from all `q`s), a tensor of shape `(batch_size, num_heads, seqlen_k)`.

While `log_normalizer` is a free lunch, which is required to compute the attention output, `total_attention` comes with the cost of launching another kernel, with a different grid partitioning. 

图片

example usage:

```python
import torch
import flag_attn

B, H, M, N, D = 2, 4, 100, 100, 128
causal = False
q = torch.randn((B, H, M, D), dtype=torch.bfloat16, device="cuda")
k = torch.randn((B, H, N, D), dtype=torch.bfloat16, device="cuda")
v = torch.randn((B, H, N, D), dtype=torch.bfloat16, device="cuda")

o, log_normalizer, cum_attn = flag_attn.flash_attention(q, k, v, causal=causal, return_log_normalizer=True, return_total_attention=True)
```

### CODE REVIEW

File: README_cn.md [hunk 1]
```diff
@@ -101,9 +101,14 @@ python piecewise_benchmark.py
...
 
 ```python
-flash_attention(q, k, v, causal=False, sm_scale=None)
+flash_attention(q, k, v, causal=False, sm_scale=None, return_log_normalizer=False, return_total_attention=False)
 ```
 
+除了 attention 的输出之外，它还可以根据 `return_log_normalizer` 和 `return_total_attention=False` 返回一些额外的输出。
+
+1. log_normalizer: 形状 (batch_size, num_heads, seqlen_q), attention 运算内部的 softmax 运算的 log normalizer.
+2. total_attention: 形状 (batch_size, num_heads, seqlen_q). attention weights 沿着 q 的序列轴上求和的结果。
```
Comment (tongxin) [inline README_cn.md]: seqlen_q -> seqlen_k

File: README.md [hunk 2]
```diff
@@ -104,9 +104,14 @@ python piecewise_benchmark.py
 The implementation of FlashAttention in the Triton language. The interface is.
 
 ```python
-flash_attention(q, k, v, causal=False, sm_scale=None)
+flash_attention(q, k, v, causal=False, sm_scale=None, return_log_normalizer=False, return_total_attention=False)
```
Comment (tongxin) [inline README.md:107]: It seems more appropriate to name it `return_total_scores_by_key` ?

File: src/flag_attn/flash.py [hunk 3]
```diff
@@ -45,14 +46,38 @@ def forward(ctx, q, k, v, causal, sm_scale):
...
         ctx.causal = causal
-        return o
+
+        has_extra_return = return_log_normalizer or return_total_attention
+        if has_extra_return:
+            outs = (
+                o,
+                L if return_log_normalizer else None,
+                tot_attn if return_total_attention else None
+            )
```
Comment (tongxin) [inline src/flag_attn/flash.py:74]: what about only one extra return is needed?

File: README.md [hunk 4]
```diff
@@ -104,9 +104,14 @@ python piecewise_benchmark.py
...
 ```
 
+In addition to the attention outputs, it can return some extra outputs dependes on `return_log_normalizer` and `return_total_attention`.
+
+1. log_normalizer: shape (batch_size, num_heads, seqlen_q). The log normalizer of the softmax inside attention operation.
+2. total_attention: shape (batch_size, num_heads, seqlen_q). The sum of attention weights along q's sequence axis.
```
[the remark this answers is not in the document]
Comment (iclementine) [reply README.md]: done

File: README_cn.md (hunk 1 above)
Comment (iclementine) [reply README_cn.md]: done

File: README.md (hunk 2 above)
Comment (iclementine) [reply README.md:107]: I think it is more common that `attention score` refers to scaled `dot(q, k)`, a.k.a the logits, the unnormalized score.

Comment (iclementine) [reply README.md:107]: For example
https://slds-lmu.github.io/seminar_nlp_ss20/attention-and-self-attention-for-nlp.html
图片
图片

File: src/flag_attn/flash.py (hunk 3 above)
Comment (iclementine) [reply src/flag_attn/flash.py:74]: I want to preserve the structure, which means to return the extra slots when any of the extra_outputs is requested.

When no extra outpus are requested, only return the attention outputs.
When any extra outputs is requested, return a 3-tuple. (attention_outputs, log_normalizer, total_attention). Though the extra-outputs that are not requested is `None`.

**Comment by iclementine** [comment]
A critical file total.py was missing in thisPR. It was added in commit `bafd600315cd41086ae142dcd234fbadbf8eca9b`.
